packages/astro/src/container/index.ts

```diff
@@ -0,0 +1,468 @@
+		const config = await validateConfig(astroConfig, process.cwd(), 'container');
+		const loadedRenderers =  await Promise.all(
+			renderers.map(async (renderer) => {
+				const mod = await import(renderer.serverEntrypoint);
+				if (typeof mod.default !== 'undefined') {
+					return {
+						...renderer,
+						ssr: mod.default,
+					} as SSRLoadedRenderer;
+				}
+				return undefined;
+			})
+		);
+		const finalRenderers = loadedRenderers.filter((r): r is SSRLoadedRenderer => Boolean(r));
+		
+		return new experimental_AstroContainer({ streaming, renderers: finalRenderers, config });
+	}
+
+	// NOTE: we keep this private via TS instead via `#` so it's still available on the surface, so we can play with it.
+	/**
+	 * 
+	 * @param manifest
+	 * @private
+	 */
+	private static async createFromManifest(manifest: SSRManifest): Promise<experimental_AstroContainer> {
+		const config = await validateConfig(ASTRO_CONFIG_DEFAULTS, process.cwd(), 'container');
+		const container = new experimental_AstroContainer({
+			manifest,
+			config,
+		});
+		container.#withManifest = true;
+		return container;
+	}
+
+	/**
+	 * Use this method to manually insert a route inside the container.
+	 *
+	 * This method is useful if you're rending a component that attempt to render redirect or a rewrite.
+	 *
+	 * @param {object} options
+	 * @param {string} options.path The path of the route. It has to match what you see in a URL browser, e.g. `/blog/12334/first-post`
+	 * @param {ComponentInstance} options.componentInstance The a compile instance of the Astro component.
+	 * @param {string[]} options.params The params of the route. Use these when your route is dynamic. For a route `/blog/[id]/[...dynamic]`, the params are `["id", "...dynamic"]`
+	 * @param {RouteType} options.type The type of route.
+	 */
+	private insertRoute({
+		path,
+		route,
+		componentInstance,
+		params = {},
+		type = 'page',
+	}: {
+		path: string;
+		route: string,
+		componentInstance: ComponentInstance;
+		params?: Record<string, string | undefined>;
+		type?: RouteType;
+	}): RouteData {
+		const pathUrl = new URL(path, 'https://example.com');
+		const routeData: RouteData = this.#createRoute(pathUrl, route, params, type);
+		this.#pipeline.manifest.routes.push({
+			routeData,
+			file: '',
+			links: [],
+			styles: [],
+			scripts: [],
+		});
+		this.#pipeline.insertRoute(routeData, componentInstance);
+		return routeData;
+	}
+
+	/**
+	 * @description
+	 * It renders a component and returns the result as a string.
+	 * 
+	 * ## Example
+	 * 
+	 * ```js
+	 * import Card from "../src/components/Card.astro";
+	 * 
+	 * const container = await AstroContainer.create();
+	 * const result = await container.renderToString(Card);
+	 * 
+	 * console.log(result); // it's a string
+	 * ```
+	 * 
+	 * 
+	 * @param {AstroComponentFactory} component The instance of the component.
+	 * @param {ContainerRenderOptions=} options Possible options to pass when rendering the component.
```

Comment:
What's this `=` at the end of the type? Saw it in a few places

Reply:
With TSDoc, we should also be able to remove the type altogether so we don't have to declare it in two places.

Reply:
@Princesseuh It's a JSDoc thing to mark something as optional